lib/time/time.go

```
@@ -0,0 +1,515 @@
+func (t Time) Attr(name string) (starlark.Value, error) {
+	switch name {
+	case "year":
+		return starlark.MakeInt(time.Time(t).Year()), nil
+	case "month":
+		return starlark.MakeInt(int(time.Time(t).Month())), nil
+	case "day":
+		return starlark.MakeInt(time.Time(t).Day()), nil
+	case "hour":
+		return starlark.MakeInt(time.Time(t).Hour()), nil
+	case "minute":
+		return starlark.MakeInt(time.Time(t).Minute()), nil
+	case "second":
+		return starlark.MakeInt(time.Time(t).Second()), nil
+	case "nanosecond":
+		return starlark.MakeInt(time.Time(t).Nanosecond()), nil
+	case "unix":
+		return starlark.MakeInt64(time.Time(t).Unix()), nil
+	case "unix_nano":
+		return starlark.MakeInt64(time.Time(t).UnixNano()), nil
+	}
+	return builtinAttr(t, name, timeMethods)
+}
+
+// AttrNames lists available dot expression strings for time. required by
+// starlark.HasAttrs interface.
+func (t Time) AttrNames() []string {
+	return append(builtinAttrNames(timeMethods),
+		"year",
+		"month",
+		"day",
+		"hour",
+		"minute",
+		"second",
+		"nanosecond",
+		"unix",
+		"unix_nano",
+	)
+}
+
+// CompareSameType implements comparison of two Time values. required by
+// starlark.Comparable interface.
+func (t Time) CompareSameType(op syntax.Token, yV starlark.Value, depth int) (bool, error) {
+	x := time.Time(t)
+	y := time.Time(yV.(Time))
+	cp := 0
+	if x.Before(y) {
+		cp = -1
+	} else if x.After(y) {
+		cp = 1
+	}
+	return threeway(op, cp), nil
+}
+
+// Binary implements binary operators, which satisfies the starlark.HasBinary
+// interface
+//    time + duration = time
+//    time - duration = time
+//    time - time = duration
+func (t Time) Binary(op syntax.Token, yV starlark.Value, side starlark.Side) (starlark.Value, error) {
+	x := time.Time(t)
+
+	switch op {
+	case syntax.PLUS:
+		switch y := yV.(type) {
+		case Duration:
+			return Time(x.Add(time.Duration(y))), nil
+		case Time:
+			return nil, fmt.Errorf("cannot add %s to %s", t.Type(), yV.Type())
+		}
```

Not sure if we need to return this error here or `nil, nil` by removing this part...